reload/fs.go

```diff
@@ -0,0 +1,60 @@
+package reload
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+)
+
+type FS struct {
+	embed fs.FS
+	dir   fs.FS
+}
+
+func NewFS(embed fs.ReadDirFS, dir string) FS {
+	return FS{
+		embed: embed,
+		dir:   os.DirFS(dir),
+	}
+}
+
+func (f FS) Open(name string) (fs.File, error) {
+	if name == "embed.go" {
+		return nil, fs.ErrNotExist
+	}
+	file, err := f.getFile(name)
+	if name != "." {
+		return file, err
+	}
+	return rootFile{file}, err
+}
+
+func (f FS) getFile(name string) (fs.File, error) {
+	cfgFile := "./.buffalo.dev.yml"
+	if _, err := os.Stat(cfgFile); err != nil {
+		return f.embed.Open(name)
+	}
+	return f.dir.Open(name)
+}
+
+type rootFile struct {
+	fs.File
+}
+
+func (f rootFile) ReadDir(n int) ([]fs.DirEntry, error) {
+	dir, ok := f.File.(fs.ReadDirFile)
+	if !ok {
+		return nil, fmt.Errorf("failed at hiding embed.go file")
+	}
+	entries, err := dir.ReadDir(n)
+	if err != nil {
+		return entries, err
+	}
+
+	for i, entry := range entries {
+		if entry.Name() == "embed.go" {
+			entries = append(entries[:i], entries[i+1:]...)
+		}
```

Comment:
It is buggy that if "embed.go" is there, we will return n-1 entries.
If we cannot do this when `NewFS()` is called and construct the object, maybe we need to do `dir.ReadDir(n+1)` and drop the last entry by a certain condition... What do you think?

Reply:
It's not really a bug. The documentation to `fs.ReadDirFile` ([here](https://pkg.go.dev/io/fs#ReadDirFile)) says:

> If n > 0, ReadDir returns **at most** n DirEntry structures.
> In this case, if ReadDir returns an empty slice, it will return
> a non-nil error explaining why.
> At the end of a directory, the error is io.EOF.

> If n <= 0, ReadDir returns all the DirEntry values from the directory
> in a single slice. In this case, if ReadDir succeeds (reads all the way
> to the end of the directory), it returns the slice and a nil error.
> If it encounters an error before the end of the directory,
> ReadDir returns the DirEntry list read until that point and a non-nil error.

So the current implementation would have been fine. I still changed it, so when `n > 0` it reads `n + 1` Entries from the underlying FS and then remove `embed.go`. If for some reason `embed.go` wasn't found the last entry is removed instead to return `n` entries.

Reply:
My interpretation of **at most n** is that sometimes it could return `m` (`m`<`n`) entries if there are only `m`, fewer files than `n`. So if it just returns 5 when I call with 6, the user could consider it as there are only 5 files in the directory. I don't think this is critical for our usage but I just want to be clear with it :-)

Anyway, Thanks for your kind consideration!